File: firebase/functions/lib/community/get_community_capabilities.dart
import 'dart:async';

import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import '../on_call_function.dart';
import '../utils/subscription_plan_util.dart';
import '../utils/utils.dart';
import 'package:data_models/cloud_functions/requests.dart';

class GetCommunityCapabilities
    extends OnCallMethod<GetCommunityCapabilitiesRequest> {
  GetCommunityCapabilities()
      : super(
          'getCommunityCapabilities',
          (jsonMap) => GetCommunityCapabilitiesRequest.fromJson(jsonMap),
        );

  @override
  Future<Map<String, dynamic>> action(
    GetCommunityCapabilitiesRequest request,
    CallableContext context,
  ) async {
    orElseUnauthorized(
      context.authUid != null,
      logMessage: 'Context auth ID was null',
    );

    final capabilities = await subscriptionPlanUtil.calculateCapabilities(
      request.communityId,
      requesterUserId: context.authUid,
    );
    return capabilities.toJson();
  }
}

File: firebase/functions/src/community/get_community_capabilities.ts
import * as functions from 'firebase-functions';
import { OnCallMethod } from '../on_call_function';
import { GetCommunityCapabilitiesRequest, PlanCapabilityList } from '../types';
import { orElseUnauthorized } from '../utils/utils';
import { subscriptionPlanUtil } from '../utils/subscription_plan_util';

export class GetCommunityCapabilities extends OnCallMethod<GetCommunityCapabilitiesRequest> {
  constructor() {
    super('getCommunityCapabilities', (json) => json as GetCommunityCapabilitiesRequest);
  }

  async action(
    request: GetCommunityCapabilitiesRequest,
    context: functions.https.CallableContext
  ): Promise<PlanCapabilityList> {
    orElseUnauthorized(context.auth?.uid != null, { logMessage: 'Context auth ID was null' });

    return subscriptionPlanUtil.calculateCapabilities(request.communityId, {
      requesterUserId: context.auth!.uid,
    });
  }
}
